platform/logger.cpp

```diff
@@ -0,0 +1,273 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) && !defined(NDEBUG)
+#include <string.h>
+#include <stdlib.h>
+#include "platform/logger_internal.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_interface.h"
+#include "platform/mbed_assert.h"
+#if !defined(MBED_DISABLE_ISR_LOGGING) && !defined(MBED_ID_BASED_TRACING)
+#include "platform/CircularBuffer.h"
+#include "rtos/Thread.h"
+
+using namespace rtos;
+using namespace mbed;
+
+// Globals related to ISR logging
+static CircularBuffer <LOG_DATA_TYPE_, LOG_BUF_SIZE_> *log_buffer_;
+#endif
+
+// Globals related to time printing
+static const ticker_data_t *const log_ticker_ = get_us_ticker_data();
+static bool time_enable_ = true;
+
+// Globals related to external buffer capturing
+static LOG_DATA_TYPE_ *extern_buf_ = NULL;
+static uint32_t xbuf_count_ = 0;
+
+static void log_update_buf(const char *data, int32_t size)
+{
+    if ((NULL == extern_buf_) || (NULL == data) || (size <= 0)) {
+        return;
+    }
+    uint32_t count = 0;
+    while(size--)
+    {
+        extern_buf_[xbuf_count_++] = data[count++];
+        xbuf_count_ %= LOG_BUF_SIZE_;
+    }
+}
+
+#if !defined(MBED_DISABLE_ISR_LOGGING) && !defined(MBED_ID_BASED_TRACING)
+extern "C" void log_init()
+{
+    log_buffer_ = new CircularBuffer <LOG_DATA_TYPE_, LOG_BUF_SIZE_>;
```

Comment:
I assume it's undefined to call this init function multipletimes? In case it happens, would we just always allocate new one and loose the old memory allocated?